File: scripts/make_silhouettes.py
#!/usr/bin/env python3
"""
make_silhouettes.py
Converts Pokemon sprites → colored silhouettes (average color fill, preserving alpha).
Run: python3 scripts/make_silhouettes.py
"""

import os
import sys

try:
    from PIL import Image
    import numpy as np
except ImportError:
    print("Installing Pillow and numpy...")
    os.system(f"{sys.executable} -m pip install Pillow numpy --quiet")
    from PIL import Image
    import numpy as np

POKEMON_DIR = os.path.join(os.path.dirname(__file__), '..', 'public', 'pokemon')
OUT_DIR = POKEMON_DIR  # overwrite in-place (originals already committed to git)

files = sorted([f for f in os.listdir(POKEMON_DIR) if f.endswith('.png')])
print(f"Processing {len(files)} sprites...")

ok = 0
skip = 0
errors = []

for fname in files:
    src = os.path.join(POKEMON_DIR, fname)
    try:
        img = Image.open(src).convert('RGBA')
        data = np.array(img, dtype=np.float32)  # H x W x 4

        alpha = data[:, :, 3]
        mask = alpha > 30  # non-transparent pixels

        if mask.sum() == 0:
            skip += 1
            continue

        # Compute average color of visible pixels
        r = data[:, :, 0][mask].mean()
        g = data[:, :, 1][mask].mean()
        b = data[:, :, 2][mask].mean()

        # Build output: fill visible pixels with avg color, keep alpha
        out = np.zeros_like(data)
        out[:, :, 0] = np.where(mask, r, 0)
        out[:, :, 1] = np.where(mask, g, 0)
        out[:, :, 2] = np.where(mask, b, 0)
        out[:, :, 3] = data[:, :, 3]  # preserve original alpha

        result = Image.fromarray(out.astype(np.uint8), 'RGBA')
        result.save(src, 'PNG', optimize=True)
        ok += 1

        if ok % 50 == 0:
            print(f"  {ok}/{len(files)} done...")

    except Exception as e:
        errors.append((fname, str(e)))

print(f"\nDone: {ok} converted, {skip} skipped (empty), {len(errors)} errors")
if errors:
    for f, e in errors:
        print(f"  ERROR {f}: {e}")

File: scripts/make-silhouettes.js
#!/usr/bin/env node
/**
 * make-silhouettes.js
 * Converts Pokémon sprites → colored silhouettes using the average color of visible pixels.
 * Uses Sharp (already installed). Overwrites files in public/pokemon/.
 */

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const POKEMON_DIR = path.join(__dirname, '..', 'public', 'pokemon');

async function avgColorSilhouette(filePath) {
  const img = sharp(filePath).ensureAlpha();
  const { data, info } = await img.raw().toBuffer({ resolveWithObject: true });
  const { width, height, channels } = info; // channels = 4 (RGBA)

  let rSum = 0, gSum = 0, bSum = 0, count = 0;
  for (let i = 0; i < data.length; i += 4) {
    const a = data[i + 3];
    if (a > 30) {
      rSum += data[i];
      gSum += data[i + 1];
      bSum += data[i + 2];
      count++;
    }
  }

  if (count === 0) return null; // fully transparent

  const r = Math.round(rSum / count);
  const g = Math.round(gSum / count);
  const b = Math.round(bSum / count);

  // Build new pixel buffer: fill visible pixels with avg color, keep alpha
  const out = Buffer.alloc(data.length);
  for (let i = 0; i < data.length; i += 4) {
    const a = data[i + 3];
    if (a > 30) {
      out[i]     = r;
      out[i + 1] = g;
      out[i + 2] = b;
      out[i + 3] = a;
    }
    // transparent pixels stay 0
  }

  await sharp(out, { raw: { width, height, channels: 4 } })
    .png({ compressionLevel: 7 })
    .toFile(filePath);

  return [r, g, b];
}

(async () => {
  const files = fs.readdirSync(POKEMON_DIR).filter(f => f.endsWith('.png')).sort();
  console.log(`Processing ${files.length} sprites...\n`);

  let ok = 0, skipped = 0, errors = [];

  for (const fname of files) {
    const fp = path.join(POKEMON_DIR, fname);
    try {
      const color = await avgColorSilhouette(fp);
      if (!color) { skipped++; continue; }
      ok++;
      if (ok % 50 === 0) console.log(`  ${ok}/${files.length}...`);
    } catch (e) {
      errors.push(`${fname}: ${e.message}`);
    }
  }

  console.log(`\nDone! ${ok} silhouettes generated, ${skipped} skipped, ${errors.length} errors`);
  if (errors.length) errors.forEach(e => console.log('  ERROR:', e));
})();
